opentelemetry-journald-logs/src/lib.rs

```diff
@@ -0,0 +1,293 @@
+                    format!("{}{}", prefix, key_str)
+                } else {
+                    key_str
+                };
+                attributes.insert(attribute_key, value_str);
+            }
+        }
+
+        let message = log_data
+            .record
+            .body
+            .as_ref()
+            .map_or_else(|| "".to_string(), format_any_value);
+
+        LogEntry {
+            message,
+            attributes,
+        }
+    }
+}
+
+fn format_any_value(value: &AnyValue) -> String {
+    match value {
+        AnyValue::Int(v) => v.to_string(),
+        AnyValue::Double(v) => v.to_string(),
+        AnyValue::String(v) => v.to_string(),
+        AnyValue::Boolean(v) => v.to_string(),
+        AnyValue::Bytes(v) => format!("{:?}", v),
+        AnyValue::ListAny(values) => {
+            let elements: Vec<String> = values.iter().map(format_any_value).collect();
+            format!("[{}]", elements.join(", "))
+        }
+        AnyValue::Map(map) => {
+            let entries: Vec<String> = map
+                .iter()
+                .map(|(k, v)| format!("{}: {}", k, format_any_value(v)))
+                .collect();
+            format!("{{{}}}", entries.join(", "))
+        }
+    }
+}
+
+fn get_priority(severity: &Severity) -> i32 {
+    match severity {
+        Severity::Debug | Severity::Debug2 | Severity::Debug3 | Severity::Debug4 => 7, // debug
+        Severity::Info | Severity::Info2 | Severity::Info3 | Severity::Info4 => 6,     // info
+        Severity::Warn | Severity::Warn2 | Severity::Warn3 | Severity::Warn4 => 4,     // warning
+        Severity::Error | Severity::Error2 | Severity::Error3 | Severity::Error4 => 3, // error
+        Severity::Fatal | Severity::Fatal2 | Severity::Fatal3 | Severity::Fatal4 => 2, // critical
+        _ => 5, // notice (default)
+    }
+}
+
+#[async_trait]
+impl LogExporter for JournaldLogExporter {
+    async fn export(&mut self, batch: Vec<LogData>) -> Result<(), LogError> {
+        let mut partial_failure = false;
+        for log in batch {
+            match self.send_log_to_journald(&log) {
+                Ok(_) => (),
+                Err(e) => {
+                    if e.kind() == std::io::ErrorKind::InvalidData {
+                        partial_failure = true;
+                    } else {
+                        return Err(LogError::Other(Box::new(e)));
+                    }
+                }
+            }
+        }
+        if partial_failure {
+            return Err(LogError::from(
+                "Some log messages exceeded the size limit and were not exported",
+            ));
+        }
+        Ok(())
+    }
+}
+
+fn sanitize_field_name(name: &str) -> String {
+    name.chars()
+        .map(|c| if c == '.' { '_' } else { c })
+        .skip_while(|&c| c == '_')
+        .filter(|&c| c == '_' || c.is_ascii_alphanumeric())
+        .collect::<String>()
+        .to_uppercase()
```

Comment:
I think we could avoid an allocation related to uppercase conversion by mapping the characters to upper case before collecting them into a `String`.

```rust
name.chars()
        .map(|c| if c == '.' { '_' } else { c })
        .skip_while(|&c| c == '_')
        .filter(|&c| c == '_' || c.is_ascii_alphanumeric())
        .map(|c| c.to_ascii_uppercase())
        .collect::<String>()
```